File: src/main/java/org/padler/gradle/minify/minifier/Minifier.java
package org.padler.gradle.minify.minifier;

import org.gradle.api.GradleException;

import java.io.File;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.*;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class Minifier {

    protected MinifierOptions minifierOptions = new MinifierOptions();

    protected final Report report = new Report();

    public void minify(String srcDir, String dstDir) {
        minifyInternal(srcDir, dstDir);
        System.err.println(createReport());
        if (!report.getErrors().isEmpty())
            throw new GradleException(report.getErrors() + " Errors in " + getMinifierName());
    }

    protected void minifyInternal(String srcDir, String dstDir) {
        try (Stream<Path> filesStream = Files.list(Paths.get(srcDir)).filter(f -> !f.toString().equals(srcDir))) {
            List<Path> files = filesStream.collect(Collectors.toList());
            for (Path f : files) {
                if (f.toFile().isFile()) {
                    Path dst = Paths.get(dstDir);
                    File dstFile = new File(dst.toString(), f.getFileName().toString());
                    dstFile.getParentFile().mkdirs();
                    minify(f.toFile(), dstFile);
                } else if (f.toFile().isDirectory()) {
                    String newDstDir = dstDir + "/" + f.getFileName().toString();
                    minifyInternal(f.toString(), newDstDir);
                }
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    protected void minify(File srcFile, File dstFile) {
        try {
            minifyFile(srcFile, dstFile);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    protected abstract String getMinifierName();

    protected abstract void minifyFile(File srcFile, File dstFile) throws IOException;

    protected String createReport() {
        StringBuilder reportStr = new StringBuilder();
        for (Error error : report.getErrors()) {
            reportStr.append("Error: ");
            reportStr.append(error);
            reportStr.append("\n");
        }
        for (Warning warning : report.getWarnings()) {
            reportStr.append("Warning: ");
            reportStr.append(warning);
            reportStr.append("\n");
        }
        reportStr.append(getMinifierName())
                .append(": ")
                .append(report.getErrors().size())
                .append(" error(s), ")
                .append(report.getWarnings().size())
                .append(" warning(s)\n");
        return reportStr.toString();
    }

    protected void writeToFile(File dstFile, String string) {
        OpenOption create = StandardOpenOption.CREATE;
        OpenOption write = StandardOpenOption.WRITE;
        OpenOption truncateExisting = StandardOpenOption.TRUNCATE_EXISTING;
        try {
            Files.write(dstFile.toPath(), string.getBytes(), create, write, truncateExisting);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    public MinifierOptions getMinifierOptions() {
        return minifierOptions;
    }

    public void setMinifierOptions(MinifierOptions minifierOptions) {
        this.minifierOptions = minifierOptions;
    }
}

File: src/main/kotlin/org/padler/gradle/minify/minifier/Minifier.kt
package org.padler.gradle.minify.minifier

import org.gradle.api.GradleException
import org.padler.gradle.minify.minifier.result.Report
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.io.UncheckedIOException
import java.nio.file.*
import java.util.*
import java.util.stream.Collectors

abstract class Minifier {

    val report = Report()

    companion object {

        private val LOGGER = LoggerFactory.getLogger(Minifier::class.java)
    }

    fun minify(srcDir: File, dstDir: File) {
        minifyInternal(srcDir, dstDir)
        if (LOGGER.isErrorEnabled) {
            LOGGER.error(createReport())
        }
        if (report.errors.isNotEmpty()) throw GradleException(report.errors.toString() + " Errors in " + minifierName)
    }

    private fun minifyInternal(srcDir: File, dstDir: File) {
        try {
            srcDir.listFiles()?.forEach {
                val f = it.toPath()
                if (it.isFile) {
                    val dst = dstDir.toPath()
                    var fileName = f.fileName.toString()
                    val copy = File(dst.toString(), fileName)
                    if (!minifierOptions.originalFileNames) {
                        fileName = rename(fileName)
                    }
                    val dstFile = File(dst.toString(), fileName)
                    dstFile.parentFile.mkdirs()
                    if (fileTypeMatches(f)) {
                        if (minifierOptions.copyOriginalFile && !minifierOptions.originalFileNames) {
                            Files.copy(f, copy.toPath(), StandardCopyOption.REPLACE_EXISTING)
                        }
                        minifyFileSave(f.toFile(), dstFile)
                    }
                } else if (it.isDirectory) {
                    minifyInternal(it, File(dstDir, f.fileName.toString()))
                }
            }
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }
    }

    private fun minifyFileSave(srcFile: File, dstFile: File) {
        try {
            minifyFile(srcFile, dstFile)
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }
    }

    private fun createReport(): String {
        val reportStr = StringBuilder()
        for (error in report.errors) {
            reportStr.append("Error: ")
            reportStr.append(error)
            reportStr.append("\n")
        }
        for (warning in report.warnings) {
            reportStr.append("Warning: ")
            reportStr.append(warning)
            reportStr.append("\n")
        }
        reportStr.append(minifierName)
                .append(": ")
                .append(report.errors.size)
                .append(" error(s), ")
                .append(report.warnings.size)
                .append(" warning(s)\n")
        return reportStr.toString()
    }

    protected fun writeToFile(dstFile: File, string: String) {
        val create: OpenOption = StandardOpenOption.CREATE
        val write: OpenOption = StandardOpenOption.WRITE
        val truncateExisting: OpenOption = StandardOpenOption.TRUNCATE_EXISTING
        try {
            Files.write(dstFile.toPath(), string.toByteArray(), create, write, truncateExisting)
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }
    }

    protected fun getExtension(filename: String): String {
        return Optional.ofNullable(filename)
                .filter { f: String -> f.contains(".") }
                .map { f: String -> f.substring(filename.lastIndexOf('.') + 1) }
                .orElse("")
    }

    protected abstract fun fileTypeMatches(f: Path): Boolean
    abstract val minifierOptions: MinifierOptions
    protected abstract val minifierName: String
    protected abstract fun minifyFile(srcFile: File, dstFile: File)

    protected abstract fun rename(oldName: String): String
}
